### DISCUSSION TITLE
Non observable object

### DISCUSSION THREAD

**Discussion by maccman** [xaviergonz/mobx-keystone#322] [Q&A] [answered]
Quick question - is there anyway to ensure a specific prop is not returned as as a deep observable object?

We have a `document` on a model which is an arbitrary object. We want mobx to register when we use this prop, but we don't need a deeply observed object. 

We have been using mobx's toJS on this object but we take quite a performance hit.

```
interface ProsemirrorJson {
  [key: string]: any
}

@model('Note')
export class Note extends Model({
  subject: prop<string>(() => ''),
  document: prop<ProsemirrorJson | undefined>(),
})

isObservableObject(note.document) # true (we want false)
```

cc @vojto / @xaviergonz

**Comment by maccman** [comment]
Maybe a transformer that puts it in a mobx box?

**Comment by xaviergonz** [comment, marked as answer]
do you need to to be serializable? if not then maybe put it outside the model props

```
@model('Note')
export class Note extends Model({
  subject: prop<string>(() => ''),
}) {
  document?: ProsemirrorJson;

  // or if you want it to be shallowly observable
  @observable.shallow
  document?: ProsemirrorJson;
}
```

alternatively, if you need it to be a model prop and the object is immutable you might use 'frozen'
https://mobx-keystone.js.org/frozen

**Reply by maccman** [reply]
Yes, we want it serialized. Oh - frozen looks like it'll work. How can I set a prop up so that it's always frozen?

**Reply by xaviergonz** [reply]
```
class M extends Model({
  frozenProp: prop<Frozen<number or whatever>>(),
  // or
  frozenProp: tProp(types.frozen(types.number or whatever, like types.unchecked<any>()))
}) {
  setFrozenProp(val: whatever) {
    this.frozenProp = frozen(val)
  }
}
```

**Reply by vojto** [reply]
Would it make sense to implement transformation that would freeze/unfreeze automatically?

Something like this:


```typescript
const _frozenTransform: ModelPropTransform<Frozen<ProsemirrorJson>, ProsemirrorJson> = {
  transform({originalValue}) {
    return originalValue.data
  },

  untransform({transformedValue}) {
    return frozen(transformedValue)
  },
}
```

The goal is to keep same format that comes out of `getSnapshot()`

**Edit:** Transform works nicely for storing frozen internally while exposing getters/setters, but `getSnapshot` still returns the internal structure. (So it includes frozen.)

**Reply by xaviergonz** [reply]
I think you might be able to do that using the new `fromSnapshotProcessor` / `toSnapshotProcessor` added in 0.63.0

**Reply by xaviergonz** [reply]
And yes, that transformation would make sense if that's what you want

**Reply by vojto** [reply]
Thank you, didn't know about this API. Seems like exactly what we've been looking for.
